packages/ra-ui-materialui/src/layout/Theme/ThemeProvider.tsx

```diff
@@ -0,0 +1,56 @@
+import * as React from 'react';
+import { ReactNode, useCallback, useEffect, useRef, useState } from 'react';
+
+import {
+    createTheme,
+    ThemeProvider as MuiThemeProvider,
+    Theme,
+} from '@mui/material/styles';
+import { ThemeOptions } from '@mui/material';
+import { ThemeContext, ThemeSetter } from './ThemeContext';
+
+/**
+ * This components sets up a React context that provides a function to set the
+ * Material-UI theme. It also wraps the children in a Material-UI ThemeProvider.
+ * @param props
+ * @param props.children The children of the component.
+ * @param props.theme The initial theme.
+ */
+export const ThemeProvider = ({
+    children,
+    theme: themeOverride,
+}: ThemeProviderProps) => {
+    const [theme, setTheme] = useTheme(themeOverride);
+
+    return (
+        <ThemeContext.Provider value={setTheme}>
+            <MuiThemeProvider theme={theme}>{children}</MuiThemeProvider>
+        </ThemeContext.Provider>
+    );
+};
+
+export interface ThemeProviderProps {
+    children: ReactNode;
+    theme: ThemeOptions;
+}
+
+const useTheme = (themeOverride?: ThemeOptions): [Theme, ThemeSetter] => {
```

Comment:
This one should be in its own file, with an example usage in jsDoc

Reply:
We don't even export it. It's only used internally by the `ThemeProvider`

Reply:
I think I'm confused because you didn't document how to read and set the theme ;)